### PR TITLE
[TOPI][AlterOpLayout][ARM] Enabling NHWC to NCHW layout transformation.

### PR DESCRIPTION
**Pull request by anijain2305**
@jackwish @yzhliu @FrozenGene @merrymercy 

Enables NHWC to NCHW conversion for ARM. Follows AlterOpLayout principle to propagate the layout as deep as possible. For my toy network with 5 convs, it results in only 2 layout transforms - 1 at the beginning, and one at the end.

WIP for tests.

### CODE REVIEW

**Comment by anijain2305** [comment]
Ping for review - @jackwish @merrymercy @FrozenGene @yzhliu 

AutoTVM would not work for now, because there is inter-dependency between AlterOpLayout and Task Extraction. This PR tackles the conversion from NHWC to NCHW layout. The test example has a sequence of NHWC ops, and transformation bring 2 LT - 1 at the start and 1 at the end. Please review.

**Comment by FrozenGene** [comment]
> Ping for review - @jackwish @merrymercy @FrozenGene @yzhliu
> 
> AutoTVM would not work for now, because there is inter-dependency between AlterOpLayout and Task Extraction. This PR tackles the conversion from NHWC to NCHW layout. The test example has a sequence of NHWC ops, and transformation bring 2 LT - 1 at the start and 1 at the end. Please review.

If we set AutoTVM task extraction opt_level to 3 (AlterOpLayout), could we make AutoTVM and this PR work?

**Comment by anijain2305** [comment]
I tried to call AtlerOpLayout() before extracting the tasks, it was able to extract the NCHW convs and tune them. The problem is while applying them. I think while searching in logs, it is still using the old NHWC layout conv, and cant find one in the logs. So, fallsback to default.

**Comment by anijain2305** [comment]
@FrozenGene I made it work. The workload had to be modified. I tried with my 2 conv network. It finished tuning, and also was able to apply both direct and winograd templates. Please review.

**Comment by anijain2305** [comment]
@jackwish It will be good if you could review this as well :)

File: topi/python/topi/arm_cpu/conv2d.py [hunk 1]
```diff
@@ -508,40 +506,62 @@ def _alter_conv2d_layout_arm(attrs, inputs, tinfos, F):
...
-        # query config of this workload
-        workload = autotvm.task.args_to_workload(
-            [data, kernel, strides, padding, dilation, layout, out_dtype], conv2d)
         target = tvm.target.current_target()
         dispatch_ctx = autotvm.DispatchContext.current
         cfg = dispatch_ctx.query(target, workload)
 
         if cfg.is_fallback:  # if is fallback, clear query cache and return None
             autotvm.task.clear_fallback_cache(target, workload)
+            if layout != 'NCHW':
```
Comment (jackwish) [inline topi/python/topi/arm_cpu/conv2d.py]: What about change this to be something like https://github.com/apache/incubator-tvm/pull/4249/files#diff-7ac4fe4f71257850a5471c9c63647bb0R530 (L530)?

**Comment by anijain2305** [comment]
@jackwish I agree that the the implementation is distributed in a long function right now. The reason (that I also got to know very recently while trying to use AutoTVM) is the tight relationship between AutoTVM and AlterOpLayout. In AutoTVM, we need to first extract the right kernels to tune (so we need to run AlterOpLayout to ensure that we have right layouts). And then after tuning is finished, we want to apply those tuned config. This again now needs to go through AlterOpLayout where we set up the config this time while also changing the layouts.


Legalize and AlterOpLayout might seem very similar to each other, but they are somewhat different. Legalize is more like a local optimization where when we see a Relay operator, we can replace it with a sequence of another Relay operators. The link that you pasted for Legalize in ARM CPU was basically inserting layout transforms before and after each conv. AlterOpLayout, on the other hand, is somewhat of a global optimization, where the layouts are basically propagated as deep as possible before inserting the layout transform. Therefore, in this CR, I am removing the legalize and moving towards the AlterOpLayout. Once we insert the layout transform for first conv, it propagates it to next operator and so on.

One way to reduce all this complexity is to write another pass that basically converts the whole graph from NHWC to NCHW (irrespective of target). One can call this pass before calling relay.build. In current scenario, where ARM only supports NCHW, that pass will be extremely useful. And then the AlterOpLayout remains unchanged. However, that pass is not yet finalized, and might take some time - https://discuss.tvm.ai/t/layout-conversion-pass/4009

In current implementation, I don't know of any better way to solve the problem. @yzhliu @kevinthesun  Please let us know what you think.

**Comment by jackwish** [comment]
Thanks for your detailed explanation @anijain2305 ! I missed you change of removing legalization.

AFAIK, AlterOpLayout is not always enabled for different *opt level*, in which case the NHWC computing may encounter build failure. Not sure AlterOpLayout is a good solution for all cases.

**Comment by tqchen** [comment]
I think we want an explicit LayoutConvert pass besides the automatic one to give developer more flexibility in the future

**Comment by anijain2305** [comment]
> Thanks for your detailed explanation @anijain2305 ! I missed you change of removing legalization.
> 
> AFAIK, AlterOpLayout is not always enabled for different _opt level_, in which case the NHWC computing may encounter build failure. Not sure AlterOpLayout is a good solution for all cases.

That is a good point. But, even Legalize is enabled at opt_level >= 1. AlterOpLayout is enabled at opt_level = 3. So, I think this is ok. As tq said, I think we should work on the Explicit layout Convertor Pass to solve the opt_level problem separately. I have some thoughts on that and take that task up.

**Comment by jackwish** [comment]
That sounds good, thank you :) @anijain2305

**Comment by yzhliu** [comment]
Thanks @anijain2305 @FrozenGene @jackwish @tqchen

**Comment by jackwish** [comment]
I have recently updated the NHWC schedule, and got comparable performance on Rasp 3B+ (https://github.com/apache/incubator-tvm/pull/3859#issuecomment-556860827). I guess this can be reverted to let the NHWC has its default schedule? @anijain2305

**Pull request merged**